keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2164 @@
+controller spec references a ResourceClaim by name, all pods created by that
+controller also use that name and thus share the resources allocated for that
+ResourceClaim.
+
+But often, each Pod is meant to have exclusive access to its own ResourceClaim
+instance instead. To support such ephemeral resources without having to modify
+all controllers that create Pods, an entry in the new PodSpec.ResourceClaims
+list can also be a ResourceClaimTemplate. When a Pod gets created, such a
+template will be used to create a normal ResourceClaim with the Pod as owner
+with an
+[OwnerReference](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#OwnerReference)),
+and then the normal allocation of the resource takes place. Once the pod got
+deleted, the Kubernetes garbage collector will also delete the
+ResourceClaim.
+
+The difference between persistent and ephemeral resources for kube-scheduler
+and kubelet is that the name of the ResourceClaim needs to be determined
+differently: the name of an ephemeral ResourceClaim is `<pod>-<resource name in
+the pod spec>`. Ownership must be checked to detect accidental conflicts with
+persistent ResourceClaims or previous incarnations of the same ephemeral
+resource. This is the same approach that was chosen for [generic ephemeral
+volumes](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1698-generic-ephemeral-volumes/README.md).
+For a resource driver there is no difference.
+
+### Coordinating resource allocation through the scheduler
+
+For immediate allocation, scheduling Pods is simple because the
+resource is already allocated and determines the nodes on which the
+Pod may run. The downside is that pod scheduling is less flexible.
+
+For delayed allocation, a node is selected tentatively by the scheduler
+and driver(s) try to allocate their resources for that node. If that
+succeeds, the Pod can get scheduled. If it fails, the scheduler must
+determine whether some other node fits the requirements and if so,
+request allocation again. If no node fits because some resources were
+already allocated for a node and are only usable there, then those
+resources must be released and then get allocated elsewhere.
+
+This is a summary of the necessary [kube-scheduler changes](#kube-scheduler) in
+pseudo-code:
+
+```
+while <pod needs to be scheduled> {
+  <choose a node, considering potential availability for those resources
+   which are not allocated yet and the hard constraints for those which are>
+  if <no node fits the pod> {
+    if <at least one resource
+            is allocated and unused,
+            uses delayed allocation, and
+            was not available on a node> {
+      <randomly pick one of those resources and tell resource driver to free it>
+    }
+  } else if <all resources allocated> {
+    <schedule pod onto node>
+  } else if <some unallocated resource uses delayed allocation> {
+    <tell resource driver to allocate for the chosen node>
```

Comment:
I can't tell whether this proposal would work, but I have some worries that I don't have with the current KEP because that has been prototyped based on the result of the 1.24 review period (i.e. without the recent optimizations, but those feel safe).

For example:

> When all entities have offered suggestions, the scheduler intersects them and either has a decision or needs to ask the entities to try again.

At what what point would the scheduler know that enough information has been gathered to make a decision? The "collaborative" part implies that the involved parties could change their information in response to information published by others, so it's not clear when the current state is stable.

We could use a time bound ("you have 10 seconds, then I'll decide"),  or use a smaller "no changes for 5 seconds". But either way, Pod scheduling always would get delayed by a certain amount.

With the current proposal, the scheduler can start a new attempt whenever something changes.

> If the decision turns out to not work, we start over (optimistic concurrency)

But then we still need to undo some work done for the previous attempt, which has the same complications (when does someone decide what to undo?).

> If we're going to do coordinated scheduling, let's do it as a primary concept?

I'm not sure. This sounds like it'll be even more complex and the benefits are not clear to me. Is the expectation that volume provisioning also uses this new primary concept?

The current delayed volume provisioning protocol is part of the GA Kubernetes API. CSI driver deployments will continue to use it, so the scheduler would have to continue to have support for it.

Reply:
I can see a path from the current proposal (one `ResourceClaimSchedulering` object per `ResourceClaim`) towards a more complex `PodScheduling` object that has a single `PotentialNodes` field and multiple entries for every pending `ResourceClaim`.

When ignoring other potential semantic changes (like whether one resource driver pays attention to what others are doing) that would be equivalent to the current proposal. An immediate advantage is that everything is stored in one place and `PotentialNodes` only gets stored only once per pod. The downside is that the object becomes larger and different, unrelated drivers need to watch it.

Filtering by label isn't possible because a single `resource-driver.k8s.io` label as in the current proposal for `ResourceClaimSchedulering` cannot hold multiple different driver names. I'm not sure whether can use the driver name as part of a label key instead of its value.

That only matters when a pod has more than one pending resource. For a single resource it's and single resource driver per cluster it's pretty much the same.

Reply:
Embedding the name into the label key is generally frowned upon.  I'd suggest we not start there, and have a conversation about what other sorts of filtering we can do with, e.g. field selectors.

Reply:
Agreed, let's not worry about that now. In the current revision I didn't include any mechanism for server-side filtering. Only scheduler and resource driver controllers need to watch PodScheduling objects, so sending everything to everyone isn't too bad.